packages/grafana-ui/src/types/panelOptions.ts

```diff
@@ -0,0 +1,192 @@
+import { SelectOptionItem } from '../components/Select/Select';
+
+/**
+ * Option editors, i.e. BooleanOption have to comply to OptionInputAPI.
+ * OptionInputAPI represents UI elements that allow value modification, name it string, boolean, complex object, whatever
+ */
+
+interface OptionInputConfig {
+  required: boolean;
+  label?: string;
+}
+export interface OptionInputAPI<T> {
+  properties?: OptionInputConfig;
+  value: T;
+  onChange: (value: T, event?: React.SyntheticEvent<HTMLElement>) => void;
+}
+
+export interface OptionsRowModel<TOptions> {
+  columns: number;
+  content: Array<OptionUIModel<TOptions, any> | GroupLayoutUIModel<any>>;
+}
+
+export enum OptionsUIType {
+  Layout = 'options:layout',
+  Group = 'options:group',
+  Editor = 'options:editor',
+}
+
+export interface OptionsUI<T> {
+  type: T;
+}
+export interface OptionsGrid extends OptionsUI<OptionsUIType.Layout> {
```

Comment:
FYI I'm going to get rid of `OptionGrid`. So the UI model could accept groups and editors only. I'm thinking about Layout being a specific type of group that could be configured using OptionsGroup model, especially it accepts `component` that is used to render a group. 

If we want to implement i.e. PanelOptionsGrid layout using this API then we will add `.addPanelOptionsGrid` higher order method to `OptionsUIBuilder`, that will use `addGroup` underneath.